docs/architecture/adr-011-generalize-genesis-accounts.md

```diff
@@ -0,0 +1,148 @@
+type GenesisState struct {
+	Params   Params             `json:"params" yaml:"params"`
+	Accounts []exported.Account `json:"accounts" yaml:"accounts"`
+}
+```
+`auth.InitGenesis` and `auth.ExportGenesis` (un)marshal accounts as well as params.
+```go
+// InitGenesis - Init store state from genesis data
+func InitGenesis(ctx sdk.Context, ak AccountKeeper, data GenesisState) {
+	ak.SetParams(ctx, data.Params)
+	// load the accounts
+	for _, a := range data.Accounts {
+		acc := ak.NewAccount(ctx, a) // set account number
+		ak.SetAccount(ctx, acc)
+	}
+}
+
+// ExportGenesis returns a GenesisState for a given context and keeper
+func ExportGenesis(ctx sdk.Context, ak AccountKeeper) GenesisState {
+	params := ak.GetParams(ctx)
+	accounts := ak.GetAllAccounts(ctx)
+	return NewGenesisState(params, accounts)
+}
+```
+
+#### 2) registering custom account types on the auth codec
+
+The auth codec must have all custom account types registered to marshal them. We will follow the pattern established in gov for proposals.
+
+A custom account definition:
+```go
+import authTypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+// Register the module account type with the auth module codec so it can decode module accounts stored in a genesis file
+func init() {
+	authtypes.RegisterAccountTypeCodec(ModuleAccount{}, "cosmos-sdk/ModuleAccount")
+}
+type ModuleAccount struct {
+    ...
+```
+The auth codec definition:
+```go
+var ModuleCdc *codec.Codec
+func init() {
+    ModuleCdc = codec.New()
+    // register module msg's and Account interface
+    ...
+	// leave the codec unsealed
+}
+// RegisterAccountTypeCodec registers an external account type defined in another module for the internal ModuleCdc.
+func RegisterAccountTypeCodec(o interface{}, name string) {
+	ModuleCdc.RegisterConcrete(o, name, nil)
+}
+```
+
+#### 3) genesis validation for custom account types
+Modules implement a `ValidateGenesis` method. As auth does not know of account implementations, accounts will need to validate themselves.
+
+We will add a `Validate` method to the `auth.Account` interface.
+
+Then auth.ValidateGenesis becomes:
+```go
+// ValidateGenesis performs basic validation of auth genesis data returning an
+// error for any failed validation criteria.
+func ValidateGenesis(data GenesisState) error {
+	// Validate params
+	...
+
+	// Validate accounts
+	addrMap := make(map[string]bool, len(data.Accounts))
+	for _, acc := range data.Accounts {
+
+		// check for duplicated accounts
+		addrStr := acc.GetAddress().String()
+		if _, ok := addrMap[addrStr]; ok {
+			return fmt.Errorf("duplicate account found in genesis state; address: %s", addrStr)
+		}
+		addrMap[addrStr] = true
+
+		// check account specific validation
+		if err := acc.Validate(); err != nil {
+			return fmt.Errorf("invalid account found in genesis state; address: %s, error: %s", addrStr, err.Error())
+		}
+
+	}
+	return nil
+}
+```
+
+#### 4) move add-genesis-account cli to auth
```

Comment:
I would remove this command entirely. Since this now entirely depends on the app, let the app define and register the CLI command(s).

Reply:
We need to make sure that apps have this functionality somehow tho. They need to be able to generate dummy accounts in genesis for testing.

Reply:
We used this command frequently during local development and testnet setup. I'd be in favour of keeping it (maybe simplify to only add baseAccounts), or adding the functionality somewhere else in the flow for setting up a new chain via cli.